program/src/processor/slash.rs

```diff
@@ -11,14 +24,187 @@ use crate::instruction::accounts::{Context, SlashAccounts};
+
+    require!(
+        ctx.accounts.stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "stake"
+    );
+
+    let mut stake_data = ctx.accounts.stake.try_borrow_mut_data()?;
+    let stake = bytemuck::try_from_bytes_mut::<Stake>(&mut stake_data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        stake.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "stake",
+    );
+
+    let (derivation, _) =
+        find_stake_pda(&stake.validator_vote, ctx.accounts.config.key, program_id);
+
+    require!(
+        ctx.accounts.stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "stake",
+    );
+
+    // slash authority
+    // - must be a signer
+    // - must match the slash authority on the config account
+    //
+    // When there is no slash authority set, the stake account cannot be slashed and
+    // an error is returned.
+
+    let slash_authority = config.slash_authority.into();
+
+    if let Some(slash_authority) = slash_authority {
+        require!(
+            ctx.accounts.slash_authority.key == &slash_authority,
+            StakeError::InvalidAuthority,
+            "slash authority",
+        );
+    } else {
+        return err!(StakeError::AuthorityNotSet, "slash authority");
+    }
+
+    require!(
+        ctx.accounts.slash_authority.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "stake authority",
+    );
+
+    // vault authority
+    // - derivation must match
+
+    let signer_bump = [config.vault_authority_bump];
+    let derivation = create_vault_pda(ctx.accounts.config.key, &signer_bump, program_id)?;
+
+    require!(
+        ctx.accounts.vault_authority.key == &derivation,
+        StakeError::InvalidAuthority,
+        "vault authority",
+    );
+
+    let signer_seeds = get_vault_pda_signer_seeds(ctx.accounts.config.key, &signer_bump);
+
+    // vault
+    // - must be the token account on the stake config account
+
+    require!(
+        ctx.accounts.vault.key == &config.vault,
+        StakeError::IncorrectVaultAccount,
+    );
+
+    let vault_data = ctx.accounts.vault.try_borrow_data()?;
+    // unpack to validate the mint
+    let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
+
+    // mint
+    // - must match the stake vault mint
+
+    require!(
+        &vault.base.mint == ctx.accounts.mint.key,
+        StakeError::InvalidMint,
+        "mint"
+    );
+
+    // Update the stake amount on both stake and config accounts.
+
+    require!(
+        stake.amount >= amount,
+        StakeError::InsufficientStakeAmount,
+        "stake"
+    );
+
+    stake.amount = stake
+        .amount
+        .checked_sub(amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    require!(
+        config.token_amount_delegated >= amount,
+        StakeError::InsufficientStakeAmount,
+        "config"
+    );
```

Comment:
I left this one generating an error, since this seems critical. As you mentioned, this should never happen. What do you think?

Reply:
Yeah, an error definitely makes sense. I wonder if this error is confusing though, since it's designed for throwing when the staker's amount is too little. Should we introduce a new one for this case?

Reply:
Good point!